management/operations.go

```diff
@@ -2,67 +2,78 @@ package management
 	ID             string
-	Status         string
+	Status         OperationStatus
 	HttpStatusCode string
-	Error          AzureError
+	Error          *AzureError
 }
 
-//getOperationStatus gets an operation given the operation ID.
-func (client *Client) getOperationStatus(operationId string) (*operation, error) {
+type OperationStatus string
+
+const (
+	OperationStatusInProgress OperationStatus = "InProgress"
+	OperationStatusSucceeded  OperationStatus = "Succeeded"
+	OperationStatusFailed     OperationStatus = "Failed"
+)
+
+// OperationId is assigned by Azure API and can be used to look up the status of an operation
+type OperationId string
+
+// getOperationStatus gets an operation given the operation ID.
+func (client *Client) GetOperationStatus(operationId OperationId) (GetOperationStatusResponse, error) {
+	operation := GetOperationStatusResponse{}
 	if operationId == "" {
-		return nil, fmt.Errorf(errParamNotSpecified, "operationId")
+		return operation, fmt.Errorf(errParamNotSpecified, "operationId")
 	}
 
-	operation := new(operation)
-	url := "operations/" + operationId
+	url := fmt.Sprintf("operations/%s", operationId)
 	response, azureErr := client.SendAzureGetRequest(url)
 	if azureErr != nil {
-		return nil, azureErr
+		return operation, azureErr
 	}
 
-	err := xml.Unmarshal(response, operation)
+	err := xml.Unmarshal(response, &operation)
 	if err != nil {
-		return nil, err
+		return operation, err
 	}
 
 	return operation, nil
 }
 
-//waitAsyncOperation blocks until the operation with the given operationId is
-//no longer in the InProgress state. If the operation was successful, nothing is
-//returned, otherwise an error is returned.
-func (client *Client) WaitAsyncOperation(operationId string) error {
-	if operationId == "" {
-		return fmt.Errorf(errParamNotSpecified, "operationId")
-	}
+const pollInterval = 30 * time.Second
 
-	status := "InProgress"
-	operation := new(operation)
-	err := errors.New("")
-	for status == "InProgress" {
-		time.Sleep(2000 * time.Millisecond)
-		operation, err = client.getOperationStatus(operationId)
+// WaitAsyncOperation blocks until the operation with the given operationId is
+// no longer in the InProgress state. If the operation was successful, nothing is
+// returned, otherwise an error is returned.
+func (client *Client) WaitAsyncOperation(operationId OperationId) error {
+	for {
+		operation, err := client.GetOperationStatus(operationId)
 		if err != nil {
-			return err
+			return fmt.Errorf("Failed to get operation status '%s': %v", operationId, err)
 		}
 
-		status = operation.Status
-	}
-
-	if status == "Failed" {
-		return fmt.Errorf("Azure operation %s failed. Code: %s, Message: %s", operationId, operation.Error.Code, operation.Error.Message)
+		switch operation.Status {
+		case OperationStatusSucceeded:
+			return nil
+		case OperationStatusFailed:
+			if operation.Error != nil {
+				return operation.Error
```

Comment:
This will allow us to do retries of async ops without doing string parsing to see what the Code was